coralproject/talk
# [CORL-2504]: implement virtualized rendering

## What does this PR do?

These changes integrate `react-virtuoso` into the comments stream and update the comment stream to use virtualized rendering.

As part of that, these changes support the Z key and the alternate oldest view. They also update how media embeds are shown (default size that can then be expanded to be viewed in full).

Also adds a new config option to the admin that controls whether comments are shown in infinite scroll or with one `Load all comments` button after the initial 20 comments.

These changes also update a few test files to use RTL.

## What changes to the GraphQL/Database Schema does this PR introduce?

This adds `loadAllComments` to the schema, to support the configuration of whether comments should load in infinite scroll or whether a Load all comments button should be shown first.

## Does this PR introduce any new environment variables or feature flags? 

no

## If any indexes were added, were they added to `INDEXES.md`?

n/a

## How do I test this PR?

Check out and interact with the stream. Add new comments, replies, change sort order, etc.

Test out Z key with Z_KEY and SEEN_COMMENTS enabled. Test Z key to root comments, replies, looping around from the end of the stream back to the beginning, with new comments via subscription, with new replies via subscription, with the Show more replies button when a comment has many replies, from a null current comment with traversal focus.

When you scroll comments up and out of view, they should be marked as seen with Z_KEY enabled. The unmark all button / shift + A should mark all comments on the story as seen, instead of just all of those in view.

Test out alternate oldest view and clicking `Add comment` takes you to the post comment form below the `Load all comments` button (or initial comments if fewer than 20). Test new comments via subscription in alternate oldest view.

Test out comments with gifs, Tweets, Youtube media, and embedded images. Test out the new expansion for this embedded media. Test out hiding/showing it.

Test out the new config in the admin around whether to show comments in infinite scroll or with a Load all comments button.
 
## How do we deploy this PR?

This requires that the story tree backend work is already deployed and story trees generated, for the Z key portion of these changes.
